socs/db/suprsync.py

```diff
@@ -208,6 +212,70 @@ def __init__(self, db_path, create_all=True, echo=False):
         if create_all:
             Base.metadata.create_all(self._engine)
 
+    def get_archive_stats(self, session=None):
+        if session is None:
+            session = self.Session()
+
+        archive_names = session.query(SupRsyncFile.archive_name)\
+            .distinct().all()
+
+        stats = {}
+        for name, in archive_names:
+            files = session.query(SupRsyncFile).filter(
+                SupRsyncFile.archive_name == name,
+            ).order_by(asc(SupRsyncFile.timestamp)).all()
+
+            finalized_until = None
+            num_files_to_copy = 0
+            last_file_added = ''
+            last_file_copied = ''
+
+            for f in files:
+                last_file_added = f.local_path
+                if (f.local_md5sum == f.remote_md5sum):
+                    last_file_copied = f.local_path
+
+                if (not f.ignore) and (f.local_md5sum != f.remote_md5sum):
+                    num_files_to_copy += 1
+
+                if finalized_until is None and not (f.ignore):
+                    if f.local_md5sum != f.remote_md5sum:
+                        finalized_until = f.timestamp - 1
+
+            # There are no more uncopied files that aren't ignored
+            if finalized_until is None:
+                finalized_until = time.time()
+
+            stats[name] = {
+                'finalized_until': finalized_until,
+                'num_files': len(files),
+                'uncopied_files': len([f for f in files if f.copied is None]),
+                'last_file_added': last_file_added,
+                'last_file_copied': last_file_copied,
+            }
+
+        return stats
+
+    def get_finalized_until(self, archive_name, session=None):
+        """
+        Returns a timetamp for which all files preceding are either successfully
+        copied, or ignored. If all files are copied, returns the current time.
+        """
+        if session is None:
+            session = self.Session()
+
+        query = session.query(SupRsyncFile).filter(
+            SupRsyncFile.archive_name == archive_name,
+        ).order_by(asc(SupRsyncFile.timestamp))
+
+        for file in query.all():
+            if file.ignore:
+                continue
+            if file.local_md5sum != file.remote_md5sum:
+                return file.timestamp - 1
```

Comment:
How close are timestamps between files generally? Is there any chance a "finalized" file has a timestamp that falls within the one second before the first failed file?

You could track the latest "good" timestamp and return that value, though then code that uses this needs to know that the timestamp means all files up to `t` are good. It seems like right now if a file has a matching timestamp it would mean it isn't finalized.